### Testing accuracy of RF classifier by providing training data of 10Hz defects and testing out it's accuracy on different speed readings of 15Hz, 20Hz, 30Hz and 40Hz. 

### Training Dataset consists of 10000 rows for each fault and testing data consists of 20000 rows for each fault. Mix data consists of 10000 rows data for each fault in all 7 rotational speeds.

In [10]:
import pandas as pd

In [11]:
data= pd.read_csv(r'D:\Acads\BTP\Train\Train10Hz.csv')

In [12]:
data.head

<bound method NDFrame.head of        X_Value  Acceleration  Acceleration_0  Acceleration_1   Voltage  \
0      0.00000     -0.768912        0.590422       -1.190895  0.004234   
1      0.00005     -1.456453        0.116244       -1.325922  0.001663   
2      0.00010     -0.488512        1.316790       -0.074655  0.002593   
3      0.00015     -1.360815        0.575074       -0.073920  0.000438   
4      0.00020     -0.395823        1.919520        0.047996  0.000992   
...        ...           ...             ...             ...       ...   
49995  0.49975     -0.221633        0.671580       -1.863335 -0.190914   
49996  0.49980      0.379733        1.455287       -1.525031 -0.196483   
49997  0.49985     -0.020645        1.169577       -1.967974 -0.190973   
49998  0.49990      0.424234        1.671135       -1.850714 -0.196692   
49999  0.49995      0.028281        1.099469       -2.238151 -0.191159   

       Voltage_0  Voltage_1  Voltage_2  Voltage_3  Identity  
0      -0.178414   

In [13]:
#shuffling train data
data= data.sample(frac=1)
data.head

<bound method NDFrame.head of        X_Value  Acceleration  Acceleration_0  Acceleration_1   Voltage  \
4086   0.20430     -0.810708       -2.835888        0.521816 -0.071273   
21828  0.09140     -0.155252        0.056081        0.549508 -0.067287   
32771  0.13855      0.327735        0.136257       -0.303540 -0.209533   
8172   0.40860      1.871966       -1.273876        1.348152 -0.152265   
18183  0.40915     -0.653605       -0.743342       -2.673374  0.188284   
...        ...           ...             ...             ...       ...   
49534  0.47670     -0.646844       -0.584955        0.318295 -0.046723   
5294   0.26470     -6.151717       -5.590221        0.593741  0.178256   
49777  0.48885     -0.038592       -0.313856       -0.482432 -0.158198   
24589  0.22945     -0.637010       -0.354864        0.215984 -0.052326   
26829  0.34145     -2.803014       -0.368493       -1.310728  0.082813   

       Voltage_0  Voltage_1  Voltage_2  Voltage_3  Identity  
4086   -0.130380   

In [14]:
import sklearn as sk

In [15]:
test_15= pd.read_csv(r'D:\Acads\BTP\Train\Train15Hz.csv')

In [16]:
test_15.head

<bound method NDFrame.head of        X_Value  Acceleration  Acceleration_0  Acceleration_1   Voltage  \
0      0.00000     -1.525662       -2.224441       -0.555582 -0.021681   
1      0.00005     -1.392776       -2.294180       -0.066691 -0.019098   
2      0.00010     -0.174552       -1.000812        0.584061 -0.019375   
3      0.00015     -0.383162       -1.101492        1.278678 -0.016724   
4      0.00020      1.119027        0.109123        2.087247 -0.017092   
...        ...           ...             ...             ...       ...   
99949  0.99925     -0.086289        1.216725       -1.050722 -0.210134   
99950  0.99930      0.034796        1.739032       -1.707600 -0.214406   
99951  0.99935     -0.306700        1.865864       -2.614070 -0.209294   
99952  0.99940     -0.349602        2.158573       -2.790634 -0.213611   
99953  0.99945     -0.223231        2.109338       -2.928357 -0.208421   

       Voltage_0  Voltage_1  Voltage_2  Voltage_3  Identity  
0       0.204297  -

In [10]:
#shuffling train data
test_15= test_15.sample(frac=1)
test_15.head

<bound method NDFrame.head of        X_Value  Acceleration  Acceleration_0  Acceleration_1   Voltage  \
68296  0.41620     -1.652402       -1.632763       -0.671985  0.217661   
54083  0.70510      0.506227       -1.259633       -0.106023 -0.102347   
75465  0.77465      0.054218        1.721106       -0.124402 -0.199095   
1246   0.06230     -3.150780        5.621717       -1.181093 -0.073666   
74318  0.71730     -0.687533       -0.622157       -0.647968 -0.203980   
...        ...           ...             ...             ...       ...   
35650  0.78295     -6.145079       -3.659867       -1.327392 -0.212658   
44494  0.22565     -1.608270       -1.877709       -2.675212  0.124784   
79819  0.99235     -4.219031       -3.270653       -1.070449 -0.076959   
80055  0.00455      0.601251        2.989919       -1.226306 -0.197102   
91148  0.55920     -0.773461        0.929050       -0.728716  0.195973   

       Voltage_0  Voltage_1  Voltage_2  Voltage_3  Identity  
68296  -0.134723  -

In [18]:
#Assigning X and Y parameter for training

dataset = data.values
X = dataset[:,0:9]
print(X)

y= dataset[:,9]
print(y)

[[ 2.043000e-01 -8.107080e-01 -2.835888e+00 ...  2.030530e-01
  -2.430000e-04 -2.887400e-02]
 [ 9.140000e-02 -1.552520e-01  5.608100e-02 ...  2.179110e-01
   3.660000e-04  3.556700e-02]
 [ 1.385500e-01  3.277350e-01  1.362570e-01 ...  1.475830e-01
   3.031000e-03  3.100500e-02]
 ...
 [ 4.888500e-01 -3.859200e-02 -3.138560e-01 ...  1.811660e-01
  -4.810000e-04 -4.566300e-02]
 [ 2.294500e-01 -6.370100e-01 -3.548640e-01 ... -1.507390e-01
  -1.290000e-06  3.940000e-03]
 [ 3.414500e-01 -2.803014e+00 -3.684930e-01 ... -2.054190e-01
  -2.680000e-05  4.712600e-02]]
[1. 0. 3. ... 4. 0. 0.]


In [19]:
#Training random forest classifier

from sklearn.ensemble import RandomForestClassifier
rf = RandomForestClassifier(n_estimators = 1000, random_state = 50)
rf.fit(X, y);

In [24]:
#Assigning X and Y parameter for testing 15Hz against 10Hz training

dataset_test_15 = test_15.values
X_test_15 = dataset_test_15[:,0:9]
print(X_test_15)

y_test_15= dataset_test_15[:,9]
print(y_test_15)

[[ 0.000000e+00 -1.525662e+00 -2.224441e+00 ... -1.836140e-01
  -1.850000e-04 -7.744000e-03]
 [ 5.000000e-05 -1.392776e+00 -2.294180e+00 ... -1.870300e-01
  -1.990000e-04 -1.546900e-02]
 [ 1.000000e-04 -1.745520e-01 -1.000812e+00 ... -1.847040e-01
  -1.530000e-04 -3.384800e-02]
 ...
 [ 9.993500e-01 -3.067000e-01  1.865864e+00 ...  2.469800e-02
  -4.310000e-04  2.185900e-02]
 [ 9.994000e-01 -3.496020e-01  2.158573e+00 ...  2.014300e-02
  -4.520000e-04 -7.977300e-02]
 [ 9.994500e-01 -2.232310e-01  2.109338e+00 ...  2.256600e-02
  -3.890000e-04 -2.033610e-01]]
[1. 1. 1. ... 4. 4. 4.]


In [25]:
#Predicting errors by sampling against test data identification result

import numpy as np
predictions_15 = rf.predict(X_test_15)
errors_15 = abs(predictions_15 - y_test_15)
print('Mean Absolute Error:', round(np.mean(errors_15), 3), 'degrees.')

Mean Absolute Error: 0.59 degrees.


In [26]:
#Calculating model accuracy through adjusting hyperparameters

accuracy = 100 - np.mean(errors_15)
print('Accuracy:', round(accuracy, 3), '%.')

Accuracy: 99.41 %.


In [27]:
#20Hz testing against 10Hz training, UR defect missing in testing
test_20= pd.read_csv(r'D:\Acads\BTP\Train\Train20Hz.csv')
test_20.head

<bound method NDFrame.head of         X_Value  Acceleration  Acceleration_0  Acceleration_1   Voltage  \
0       0.00000      1.708963       22.936699        0.418647  0.062727   
1       0.00005      0.397558       20.744087       -2.443510  0.058105   
2       0.00010      0.245741       19.886220       -4.793004  0.065566   
3       0.00015     -0.492077       16.847528       -7.492809  0.060839   
4       0.00020      1.393773       15.511186       -5.217568  0.068609   
...         ...           ...             ...             ...       ...   
100239  1.01385     -2.401899       -2.835274       -2.798354  0.168155   
100240  1.01390      1.870737       -1.357366        0.724601  0.170059   
100241  1.01395      1.843078       -2.764307        1.164236  0.166506   
100242  1.01400      5.318520        0.023913        4.574955  0.167842   
100243  1.01405      5.640224       -0.904798        2.451036  0.162821   

        Voltage_0  Voltage_1  Voltage_2  Voltage_3  Identity  
0     

In [37]:
#shuffling train data
test_20= test_20.sample(frac=1)
test_20.head

<bound method NDFrame.head of        X_Value  Acceleration  Acceleration_0  Acceleration_1   Voltage  \
99692  0.98650     -3.282805       -0.508463       -2.519232 -0.184903   
31371  0.56905      1.017242       -0.401030       -1.033813 -0.211172   
9694   0.48470     -2.368462        9.829892       -6.920231 -0.218798   
3244   0.16220      4.701049        4.572067       -2.557094  0.201480   
73353  0.66910      1.750881       -0.375860       -1.149603 -0.137631   
...        ...           ...             ...             ...       ...   
3869   0.19345     -2.368585      -10.495175        4.931882 -0.085213   
52791  0.64055     -0.962279       -1.438279       -0.843403 -0.160353   
2259   0.11295     -5.478067        0.342529       -4.296639  0.195536   
47868  0.39440     -0.740885        0.258792       -0.418840 -0.085598   
31784  0.58970      1.798701        2.474733        1.066335  0.126583   

       Voltage_0  Voltage_1  Voltage_2  Voltage_3  Identity  
99692   0.203721  -

In [38]:
#Assigning X and Y parameter for testing 20Hz against 10Hz training

dataset_test_20 = test_20.values
X_test_20 = dataset_test_20[:,0:9]
print(X_test_20)

y_test_20= dataset_test_20[:,9]
print(y_test_20)

[[ 9.865000e-01 -3.282805e+00 -5.084630e-01 ... -1.960000e-02
  -4.140000e-04  4.447800e-02]
 [ 5.690500e-01  1.017242e+00 -4.010300e-01 ...  1.848780e-01
   1.155000e-03  3.146600e-02]
 [ 4.847000e-01 -2.368462e+00  9.829892e+00 ...  7.987900e-02
  -3.130000e-04  9.424500e-02]
 ...
 [ 1.129500e-01 -5.478067e+00  3.425290e-01 ... -2.059700e-02
  -4.470000e-04 -5.449300e-02]
 [ 3.944000e-01 -7.408850e-01  2.587920e-01 ... -1.124880e-01
   6.130000e-05  2.840300e-02]
 [ 5.897000e-01  1.798701e+00  2.474733e+00 ... -2.283560e-01
  -3.640000e-04 -8.841000e-03]]
[4. 2. 1. ... 1. 0. 2.]


In [39]:
#Predicting errors by sampling against test data identification result

import numpy as np
predictions_20 = rf.predict(X_test_20)
errors_20 = abs(predictions_20 - y_test_20)
print('Mean Absolute Error:', round(np.mean(errors_20), 3), 'degrees.')

Mean Absolute Error: 0.628 degrees.


In [40]:
#Calculating model accuracy through adjusting hyperparameters

accuracy = 100 - np.mean(errors_20)
print('Accuracy:', round(accuracy, 3), '%.')

Accuracy: 99.372 %.


In [32]:
# 30Hz testing against 10Hz training
test_30= pd.read_csv(r'D:\Acads\BTP\Train\Train30Hz.csv')
test_30.head

<bound method NDFrame.head of         X_Value  Acceleration  Acceleration_0  Acceleration_1   Voltage  \
0      0.000000    -11.367953      -11.637524        6.021533 -0.211798   
1      0.000050    -12.166375       -9.140292        3.120780 -0.217642   
2      0.000100     -8.550057       -4.918858        0.881439 -0.211531   
3      0.000150     -7.510203       -1.977898       -0.244971 -0.217031   
4      0.000200     -9.059351        0.536890        3.866246 -0.210774   
...         ...           ...             ...             ...       ...   
99948  0.999350      6.308956       -7.804564        2.868125 -0.231882   
99949  0.999400      3.970360       -6.623541        1.938864 -0.229132   
99950  0.999450      1.442453       -4.509385        2.251926 -0.233108   
99951  0.999500     -0.867132       -3.078502        0.568868 -0.229825   
99952  0.999528     -0.404975       -0.450460       -0.560514  0.001299   

       Voltage_0  Voltage_1  Voltage_2  Voltage_3  Identity  
0      

In [33]:
#shuffling train data
test_30= test_30.sample(frac=1)
test_30.head

<bound method NDFrame.head of        X_Value  Acceleration  Acceleration_0  Acceleration_1   Voltage  \
52547  0.62835      2.730746       -0.476049        2.396878 -0.103508   
47987  0.40035     -5.802968        5.357739       -5.320246 -0.233981   
86163  0.31010     -4.009192        1.940024       -1.614723 -0.022028   
42095  0.10575     -2.814201        2.135122       -2.291328 -0.143554   
10342  0.51710     -0.890858       -7.986771       -6.007880  0.206436   
...        ...           ...             ...             ...       ...   
80341  0.01900      1.036296       -1.247846       -9.402180  0.206624   
5124   0.25620     -6.842208       15.403263       -2.536387  0.019909   
95117  0.75780      0.939305       -1.535889        2.138219  0.069146   
84312  0.21755     -7.991838        2.277302       -0.452413  0.189824   
97859  0.89490      2.884284       -0.916093        1.035335 -0.118711   

       Voltage_0  Voltage_1  Voltage_2  Voltage_3  Identity  
52547  -0.131813   

In [34]:
#Assigning X and Y parameter for testing 30Hz against 10Hz training

dataset_test_30 = test_30.values
X_test_30 = dataset_test_30[:,0:9]
print(X_test_30)

y_test_30= dataset_test_30[:,9]
print(y_test_30)

[[ 6.283500e-01  2.730746e+00 -4.760490e-01 ...  2.366740e-01
   7.274561e-05  8.645700e-02]
 [ 4.003500e-01 -5.802968e+00  5.357739e+00 ...  1.589240e-01
   2.670000e-04  8.835200e-02]
 [ 3.101000e-01 -4.009192e+00  1.940024e+00 ... -1.940140e-01
  -3.870000e-04  2.043420e-01]
 ...
 [ 7.578000e-01  9.393050e-01 -1.535889e+00 ...  1.420620e-01
  -4.080000e-04 -5.433000e-02]
 [ 2.175500e-01 -7.991838e+00  2.277302e+00 ... -1.606640e-01
   1.450000e-03  2.352080e-01]
 [ 8.949000e-01  2.884284e+00 -9.160930e-01 ...  2.266080e-01
  -4.610000e-04 -8.752000e-03]]
[0. 0. 4. ... 4. 4. 4.]


In [35]:
#Predicting errors by sampling against test data identification result

import numpy as np
predictions_30 = rf.predict(X_test_30)
errors_30 = abs(predictions_30 - y_test_30)
print('Mean Absolute Error:', round(np.mean(errors_30), 3), 'degrees.')

Mean Absolute Error: 0.624 degrees.


In [36]:
#Calculating model accuracy through adjusting hyperparameters

accuracy = 100 - np.mean(errors_30)
print('Accuracy:', round(accuracy, 3), '%.')

Accuracy: 99.376 %.


In [26]:
# Testing mix data from different rotational speeds of 10Hz, 15Hz, 20Hz, 25Hz, 30Hz with 200 dataset row per fault
# for each rotational speed

test_mix= pd.read_csv(r'D:\Acads\BTP\Train\Mix_test_10_15_20_25_30.csv')
test_mix.head

<bound method NDFrame.head of       X_Value  Acceleration  Acceleration_0  Acceleration_1   Voltage  \
0     0.00000     -0.768912        0.590422       -1.190895  0.004234   
1     0.00005     -1.456453        0.116244       -1.325922  0.001663   
2     0.00010     -0.488512        1.316790       -0.074655  0.002593   
3     0.00015     -1.360815        0.575074       -0.073920  0.000438   
4     0.00020     -0.395823        1.919520        0.047996  0.000992   
...       ...           ...             ...             ...       ...   
4593  0.50780      4.809103       16.989462       -4.356433  0.050899   
4594  0.50785      4.307922       14.697644       -3.574698  0.060425   
4595  0.50790      1.048958       10.596533       -6.032753  0.055586   
4596  0.50795      1.293955        9.011806       -4.070083  0.065307   
4597  0.50800     -0.179592        5.323115       -1.699513  0.059597   

      Voltage_0  Voltage_1  Voltage_2  Voltage_3  Identity  
0     -0.178414   0.174913  -0.0

In [27]:
#shuffling train data
test_mix= test_mix.sample(frac=1)
test_mix.head

<bound method NDFrame.head of       X_Value  Acceleration  Acceleration_0  Acceleration_1   Voltage  \
1989  0.34495     -3.929288       -3.510198       -2.872238  0.122159   
1355  0.29460     -3.637947       -1.462221        1.419954  0.159976   
3347  0.05740     -3.178808       -5.224212       -2.268293  0.197651   
2769  0.99395      8.319948        1.752416       -0.720016  0.208488   
3553  0.05770      4.398398       -1.238392        0.700341  0.211932   
...       ...           ...             ...             ...       ...   
2234  0.98815     -0.427170       -3.697561       -1.480554  0.230501   
4281  0.50270     12.029938        7.036883        4.376335 -0.029232   
1593  0.29610     -2.783592        0.982460       -1.678560  0.075300   
3688  0.50450      0.611823        3.032033       -8.212913 -0.202283   
1959  0.34345      0.853501        0.065658       -0.096221  0.095474   

      Voltage_0  Voltage_1  Voltage_2  Voltage_3  Identity  
1989   0.100442  -0.225641  -0.0

In [28]:
#Assigning X and Y parameter for testing 30Hz against 10Hz training

dataset_test_mix = test_mix.values
X_test_mix = dataset_test_mix[:,0:9]
print(X_test_mix)

y_test_mix= dataset_test_mix[:,9]
print(y_test_mix)

[[ 3.449500e-01 -3.929288e+00 -3.510198e+00 ... -2.256410e-01
  -4.470000e-04  1.013260e-01]
 [ 2.946000e-01 -3.637947e+00 -1.462221e+00 ... -2.245220e-01
  -5.280000e-04  3.585500e-02]
 [ 5.740000e-02 -3.178808e+00 -5.224212e+00 ... -2.020120e-01
  -1.222000e-03 -1.547790e-01]
 ...
 [ 2.961000e-01 -2.783592e+00  9.824600e-01 ...  1.633790e-01
   3.787000e-03  2.046200e-02]
 [ 5.045000e-01  6.118230e-01  3.032033e+00 ... -8.837000e-03
  -5.210000e-04  2.834300e-02]
 [ 3.434500e-01  8.535010e-01  6.565800e-02 ... -2.256040e-01
  -3.940000e-04  3.977100e-02]]
[2. 4. 3. ... 3. 4. 2.]


In [34]:
#Predicting errors by sampling against test data identification result

import numpy as np
predictions_mix = rf.predict(X_test_mix)
errors_mix = abs(predictions_mix - y_test_mix)
print('Mean Absolute Error:', round(np.mean(errors_mix), 3), 'degrees.')

Mean Absolute Error: 0.388 degrees.


In [38]:
#Calculating model accuracy through adjusting hyperparameters

accuracy = 100 - np.mean(errors_mix)
print('Accuracy:', round(accuracy, 3), '%.')

Accuracy: 99.612 %.


In [6]:
# 40Hz testing against 10Hz training
test_40= pd.read_csv(r'D:\Acads\BTP\Train\Train40Hz.csv')
test_40.head

<bound method NDFrame.head of         X_Value  Acceleration  Acceleration_0  Acceleration_1   Voltage  \
0       0.00000    -11.366601      -20.932982       -5.598388 -0.225003   
1       0.00005    -12.013575      -21.778201       -8.949557 -0.232821   
2       0.00010    -10.241557      -21.849168        0.619594 -0.227753   
3       0.00015    -11.740304      -21.493965        7.233590 -0.234780   
4       0.00020    -15.460252      -22.320645       13.823692 -0.229820   
...         ...           ...             ...             ...       ...   
49995  25.09975      1.194383        0.626765      -10.774995  0.127278   
49996  25.09980     -0.315305        0.908914      -13.720839  0.129294   
49997  25.09985     -0.943348        0.550395       -6.796599  0.123766   
49998  25.09990     -1.503535        0.413249       -1.593281  0.125808   
49999  25.09995     -3.555093        0.845560       -3.514904  0.117905   

       Voltage_0  Voltage_1  Voltage_2  Voltage_3  Comment  
0       

In [7]:
#shuffling train data
test_40= test_40.sample(frac=1)
test_40.head

<bound method NDFrame.head of         X_Value  Acceleration  Acceleration_0  Acceleration_1   Voltage  \
24176   0.20880     -3.636226        7.277042       -1.990642 -0.133710   
47631  10.08155      1.238022       -5.652961        5.024269 -0.067844   
40283   0.11415     -2.767365       -4.928066        1.188374 -0.210764   
10043   0.00215     -4.119705        5.010761       -5.197595 -0.155520   
47918  10.09590     -7.724222        3.539730        2.914073  0.168524   
...         ...           ...             ...             ...       ...   
17937   0.39685      3.218404       -2.318860       -2.220139  0.101779   
36039  25.10195      4.183641        4.493488        9.940134 -0.016434   
38647  25.23235      2.558401        0.528295        0.332876 -0.231496   
19118   0.45590    -10.707702       -8.730575       -5.674356 -0.231473   
8054    0.40270      3.191483       11.271211       -6.769398 -0.167631   

       Voltage_0  Voltage_1  Voltage_2  Voltage_3  Comment  
24176  -

In [8]:
#Assigning X and Y parameter for testing 30Hz against 10Hz training

dataset_test_40 = test_40.values
X_test_40 = dataset_test_40[:,0:9]
print(X_test_40)

y_test_40= dataset_test_40[:,9]
print(y_test_40)

[[ 2.0880000e-01 -3.6362260e+00  7.2770420e+00 ...  2.5214900e-01
   7.2700000e-05 -1.7691600e-01]
 [ 1.0081550e+01  1.2380220e+00 -5.6529610e+00 ...  2.0962900e-01
  -6.0400000e-04 -2.6760300e-01]
 [ 1.1415000e-01 -2.7673650e+00 -4.9280660e+00 ...  9.6838000e-02
  -6.7500000e-04 -9.1847000e-02]
 ...
 [ 2.5232350e+01  2.5584010e+00  5.2829500e-01 ...  1.2501000e-01
  -9.9800000e-04  4.2860000e-03]
 [ 4.5590000e-01 -1.0707702e+01 -8.7305750e+00 ...  8.1198000e-02
  -3.5200000e-04  1.4742000e-02]
 [ 4.0270000e-01  3.1914830e+00  1.1271211e+01 ... -4.5521000e-02
  -2.2600000e-04  1.5751200e-01]]
[0. 4. 4. ... 3. 2. 1.]


In [21]:
#Predicting errors by sampling against test data identification result

import numpy as np
predictions_40 = rf.predict(X_test_40)
errors_40 = abs(predictions_40 - y_test_40)
print('Mean Absolute Error:', round(np.mean(errors_40), 3), 'degrees.')

Mean Absolute Error: 0.794 degrees.


In [23]:
#Calculating model accuracy through adjusting hyperparameters

accuracy = 100 - np.mean(errors_40)
print('Accuracy:', round(accuracy, 3), '%.')

Accuracy: 99.206 %.


### Results testing different datasets against 10Hz trained model
### 15Hz : 0.59 Mean Absolute error
### 20Hz : 0.628 Mean Absolute error
### 30Hz : 0.624 Mean Absolute error
### Mix speed: 0.794 Mean Absolute error / 99.98% Accuracy